### PR TITLE
mongoose_wpool controlled restarts

### PR DESCRIPTION
**Pull request by michalwski**
This PR introduces `mongoose_wpool` restarts if the `worker_pool` was stopped.

Proposed changes:
1. All pools started via `mongoose_wpool` are organized under MongooseIM's supervisors:
    * there is a single `mongoose_wpool_sup` under which there is many
        * `mongoose_wpool_TYPE_sup` where `TYPE` is one of the known pool types. Every such a supervisor has:
            * `mongoose_wpool_TYPE_mgr` (a gen_server) started automatically. This process is responsible for starting, stopping and possibly restarting pools of the given `TYPE`.
            * The real `worker_pool` is a temporary (it never gets restarted by the supervisor) child of the `mongoose_wpool_TYPE_sup`. When the pool dies (most probably due to too many restarts), the manager restarts it.
2. Removed some supervisors which were not used.

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5792.1](https://travis-ci.org/esl/MongooseIM/jobs/448318903) / Erlang 19.3 / small_tests / a5e84540212cafc325983cd8a21ee00167ca3853
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2117/5792/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2117/5792/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-10-30_13.07.03/lib.mongooseim.logs/run.2018-10-30_13.07.03/suite.log.html)

**Comment by mongoose-im** [comment]
[5798.1](https://travis-ci.org/esl/MongooseIM/jobs/448397766) / Erlang 19.3 / small_tests / 33af8b1051239634a84d13a1ada1d71606f42885
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2117/5798/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2117/5798/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-10-30_16.09.30/lib.mongooseim.logs/run.2018-10-30_16.09.30/suite.log.html)

File: src/wpool/mongoose_wpool_mgr.erl [hunk 1]
```diff
@@ -0,0 +1,204 @@
...
+
+-spec init([any()]) -> {ok, state()}.
+init([Type]) ->
+    {ok, #state{type = Type, pools = #{}, monitors = #{}}}.
+
+-spec handle_call(request(), {pid(), term()}, state()) ->
+    {reply, reply(), state()}.
+handle_call({start_pool, Host, Tag, WpoolOpts, ConnOpts}, _From,
+            #state{type = Type, pools = Pools, monitors = Monitors} = State) ->
+    ?WARNING_MSG("Starting pool {~p,~p,~p}, with pool opts: ~p and conn opts: ~p~n",
```
Comment (fenek) [inline src/wpool/mongoose_wpool_mgr.erl]: 1. k-v logs please. :)
2. I think this one may be at info level.

File: src/wpool/mongoose_wpool_type_sup.erl [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+%%--------------------------------------------------------------------
+%% @doc
+%% Starts the supervisor
+%%
+%% @end
+%%--------------------------------------------------------------------
+-spec start_link(mongoose_wpool:type()) ->
+    {ok, Pid :: pid()} | ignore | {error, Reason :: term()}.
+start_link(Type) ->
+    io:format("what is the name :~p~n", [name(Type)]),
```
Comment (fenek) [inline src/wpool/mongoose_wpool_type_sup.erl]: Temporary debug? :)

File: src/wpool/mongoose_wpool_mgr.erl [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+            {ok, State#state{pools = NewPools, monitors = NewMonitors}};
+        Other ->
+            ?WARNING_MSG("issue=unable_to_restart_pool, pool=~p, reason=~p",
+                         [PoolKey, Other]),
+            Other
+    end.
+
+
+restart_pool(Reason, PoolKey, #state{pools = Pools} = State) ->
+    ?ERROR_MSG("event=dead_pool, name=~p, reason=~p",
```
Comment (fenek) [inline src/wpool/mongoose_wpool_mgr.erl]: 268x0w
:)

File: src/wpool/mongoose_wpool_mgr.erl [hunk 4]
```diff
@@ -0,0 +1,204 @@
...
+               [PoolKey, Reason]),
+    case maps:get(PoolKey, Pools, undefined) of
+        undefined ->
+            ?WARNING_MSG("event=unknonw_pool_dead, name=~p", [PoolKey]),
```
Comment (fenek) [inline src/wpool/mongoose_wpool_mgr.erl]: `unknown_pool_dead`

File: src/wpool/mongoose_wpool.erl [hunk 5]
```diff
@@ -50,17 +59,28 @@
...
                 {keypos, 2},
-                {heir, whereis(wpool_sup), undefined}]);
+                {heir, whereis(mongoose_wpool_sup), undefined}]);
         _ ->
             ok
     end.
 
+-spec known_types() -> [type()].
+known_types() ->
+    [redis, riak, http, rdbms, cassandra, elastic, generic].
```
[the remark this answers is not in the document]
Comment (michalwski) [reply src/wpool/mongoose_wpool.erl]: Right, I'll change it to on-demand starting.

**Comment by mongoose-im** [comment]
[5799.1](https://travis-ci.org/esl/MongooseIM/jobs/448497067) / Erlang 19.3 / small_tests / 7cb3743421bb6439e5aa5735e1bd6ddafb304316
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2117/5799/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2117/5799/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-10-30_19.12.11/lib.mongooseim.logs/run.2018-10-30_19.12.11/suite.log.html)

File: doc/developers-guide/mongoose_wpool.md [hunk 6]
```diff
@@ -0,0 +1,27 @@
+# mongoose_wpool
+
+All the [outgoing pools] configured by the `outgoing_pools` option are hidden behind `mongoose_wpool` API.
```
Comment (goddammit) [inline doc/developers-guide/mongoose_wpool.md]: All the [outgoing pools] configured by the `outgoing_pools` option are hidden behind the `mongoose_wpool` API.

File: doc/developers-guide/mongoose_wpool.md [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+Every pool is described by a tuple `{Type, Host, Tag, PoolOptions, ConnectionOptions}` (see [outgoing pools] for details about each element of the tuple).
+
+## Supervision tree
+
+![Supervision tree](mongoose_wpool_supervision.svg)
+
+* `mongoose_wpool_sup` supervisor for very type of the pool. Under it there can be many children of:
```
Comment (goddammit) [inline doc/developers-guide/mongoose_wpool.md]: `mongoose_wpool_sup` supervisor for every pool type. Under it there can be many children of:

File: doc/developers-guide/mongoose_wpool.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+    * `mongoose_wpool_type_sup` is started on-demand when a pool of given type is started.
+      Many pools of the same type are supervised by the supervisor. It's children are:
```
Comment (goddammit) [inline doc/developers-guide/mongoose_wpool.md]: Many pools of the same type are supervised by the supervisor. Its children are:

**Comment by mongoose-im** [comment]
[5801.1](https://travis-ci.org/esl/MongooseIM/jobs/448749858) / Erlang 19.3 / small_tests / 3c8f9de91b09c18386dc84927796d5f936039307
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2117/5801/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2117/5801/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-10-31_09.59.18/lib.mongooseim.logs/run.2018-10-31_09.59.19/suite.log.html)

File: doc/developers-guide/mongoose_wpool.md [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+           It's responsible for starting, stopping and restarting the pool.
+           Restarting happens when the main `worker_pool` process for the pool is stopped unintentionally.
+           This usually happens when there was too many restarts of worker processes.
+        * many `worker_pool` supervisors holding a specific pool are on the same level as the manager.
+
+The `mongoose_wpool_mgr` manages the pool by setting monitor for every started pool.
+
+## Implementing new pool type
+
+In order to add new pool type, there needs to be created module `mongoose_wpool_NEW_TYPE` implementing the `mongoose_wpool` behaviour.
```
Comment (goddammit) [inline doc/developers-guide/mongoose_wpool.md]: To add a new pool type, create a `mongoose_wpool_NEW_TYPE` module  implementing the `mongoose_wpool` behaviour.

File: doc/developers-guide/mongoose_wpool.md [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+For instance, if we are going to add new type `xyz` we need to create `mongoose_wpool_xyz` module.
```
Comment (goddammit) [inline doc/developers-guide/mongoose_wpool.md]: This means that for a new type `xyz` we need to create a `mongoose_wpool_xyz` module.

**Pull request merged**